File: app/src/main/java/cz/civilizacehra/cipherbreaker/MapDictionary.java
package cz.civilizacehra.cipherbreaker;

import android.content.Context;
import android.location.Location;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collections;

class MapDictionary extends Dictionary {

    class Point implements Comparable<Point> {
        Point(float dist, String n) {
            distance = dist;
            name = n;
        }
        float distance;
        String name;

        @Override
        public int compareTo(final Point o) {
            if (this.distance != o.distance) {
                return Float.compare(this.distance, o.distance);
            } else {
                return this.name.compareTo(o.name);
            }
        }
    }

    MapDictionary(Context context, String filename, TextView results) {
        super(context, filename, results);
        mSvjz = false;
    }

    void setSvjz(boolean svjz) {
        mSvjz = svjz;
    }

    void setLocation(Location location) {
        mLocation = location;
    }
    private Location mLocation;
    private ArrayList<Point> mSortedResults = new ArrayList<>();

    @Override
    public void findResults(String input, boolean subset, boolean exact, boolean superset, boolean hamming, boolean levenshtein, boolean regexp, int minLength, int maxLength) {

        prepare();

        setSuffix("");
        findResults_impl(input, subset, exact, superset, hamming, levenshtein, regexp, minLength, maxLength);
        if (mSvjz) {
            for (String s : mWorldSides) {
                processWithWorldSide(input, s, subset, exact, superset, hamming, levenshtein, regexp, minLength, maxLength);
            }
        }

        conclude(false);
    }

    @Override
    protected void prepare() {
        mSortedResults.clear();
        mStartTime = System.currentTimeMillis();
    }

    @Override
    protected void matched(String match) {

        String[] split = match.split(";");
        double lat = .0;
        double lon = .0;
        String name = split[0] + mSuffix;
        if (split.length >= 3) {
            lat = Double.parseDouble(split[split.length - 2]);
            lon = Double.parseDouble(split[split.length - 1]);
        }
        float distance = .0f;
        if (mLocation != null) {
            Location targetLocation = new Location("");
            targetLocation.setLatitude(lat);
            targetLocation.setLongitude(lon);
            distance = mLocation.distanceTo(targetLocation);
        }
        mSortedResults.add(new Point(distance, name));
    }

    @Override
    protected void conclude(boolean sort) {
        StringBuilder resultStr = new StringBuilder();
        int counter = 0;
        Collections.sort(mSortedResults);
        for (Point point : mSortedResults) {
            ++counter;
            String row = point.name + " (" + Math.round(point.distance) + "m)\n";
            resultStr.append(row);
            if (counter >= 3000) {
                break;
            }
        }
        mResults.setText("Result: (" + counter + ")" + computationTime() + "\n" + resultStr);
    }

    private void setSuffix(String s) {
        if (s.length() > 0) {
            mSuffix = " (" + s.toUpperCase() + ")";
        } else {
            mSuffix = "";
        }
    }

    private void processWithWorldSide(String input, String s, boolean subset, boolean exact, boolean superset, boolean hamming, boolean levenshtein, boolean regexp, int minLength, int maxLength) {
        String[] arr = s.split("");
        boolean contains = true;
        for (String c : arr) {
            if (!input.contains(c)) {
                contains = false;
            }
        }
        if (contains) {
            String modified = input;
            for (String c : arr) {
                modified = modified.replaceFirst(c, "");
            }
            setSuffix(s);
            findResults_impl(modified, subset, exact, superset, hamming, levenshtein, regexp, minLength, maxLength);
        }
    }

    private String mSuffix;
    private String[] mWorldSides = {"s", "sv", "sz", "v", "z", "j", "jv", "jz"};
    private boolean mSvjz;
}

File: app/src/main/java/cz/civilizacehra/cipherbreaker/MapDictionary.kt
package cz.civilizacehra.cipherbreaker

import android.location.Location
import java.io.InputStream
import kotlin.math.min

import kotlin.math.round

internal class MapDictionary(openDictionary: (String) -> InputStream) : Dictionary(openDictionary) {
    private var mLocation: Location? = null
    private val mSortedResults = ArrayList<Point>(2 * mMaxNumberOfResults)

    internal inner class Point(var distance: Float, var name: String) : Comparable<Point> {

        override fun compareTo(other: Point): Int {
            return if (this.distance != other.distance) {
                this.distance.compareTo(other.distance)
            } else {
                this.name.compareTo(other.name)
            }
        }
    }

    fun setLocation(location: Location) {
        mLocation = location
    }

    // Lines are name;lat;lon, names are without semicolons
    override fun name(line: String): String {
        return line.substringBefore(';')
    }

    override fun prepare() {
        mSortedResults.clear()
        mStartTime = System.currentTimeMillis()
    }

    override fun matched(match: String) {

        val split = match.split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        var lat = .0
        var lon = .0
        val name = split[0]
        if (split.size >= 3) {
            lat = split[split.size - 2].toDouble()
            lon = split[split.size - 1].toDouble()
        }
        var distance = .0f
        if (mLocation != null) {
            val targetLocation = Location("")
            targetLocation.latitude = lat
            targetLocation.longitude = lon
            distance = mLocation!!.distanceTo(targetLocation)
        }
        mSortedResults.add(Point(distance, name))
        if (mSortedResults.size >= 2 * mMaxNumberOfResults) {
            mSortedResults.sort()
            val tempList = mSortedResults.take(mMaxNumberOfResults)
            mSortedResults.clear()
            for (value in tempList) {
                mSortedResults.add(value)
            }
        }
    }

    override suspend fun conclude(): String {
        val resultStr = StringBuilder()
        var counter = 0
        mSortedResults.sort()
        for (point in mSortedResults) {
            ++counter
            val row = point.name + " (${round(point.distance).toInt()}m)\n"
            resultStr.append(row)
            if (counter >= mMaxNumberOfResults) {
                break
            }
        }
        return resultStr.toString()
    }

    override fun resultsSize(): Int {
        return min(mSortedResults.size, mMaxNumberOfResults)
    }
}
